### PR TITLE
new: driver selection in falco.yaml

### PR DESCRIPTION
**Pull request by therealbobo**
Signed-off-by: Roberto Scolaro <roberto.scolaro21@gmail.com>

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> /kind release

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area engine

> /area tests

> /area proposals

> /area CI

**What this PR does / why we need it**:

It might be useful to configure the driver mode via the configuration file `falco.yaml`

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new: driver selection in falco.yaml
```

### CODE REVIEW

**Comment by Andreagit97** [comment]
Thank you for this! @therealbobo 
We discussed many times to uniform how we open engines and this could be a possible way 
cc @leogr

**Comment by therealbobo** [comment]
I think that driver selected in `falco.yaml` should be considered if and only if the driver to be used is not specified in any other way (e.g. env variable or args). Moreover, using the config file, it will possibile to simplify the service management. 🤔

**Comment by FedeDP** [comment]
We actually need this for the shiny new falco-driver-loader integrated inside falcoctl: https://github.com/falcosecurity/falcoctl/issues/327

**Comment by FedeDP** [comment]
I'd suggest a rebase, and then:
* DEPRECATION NOTICE for `FALCO_BPF_PROBE`
* DEPRECATION NOTICE for cmd line options (note that `-u` was already dropped in the meantime, so i think that only the `modern-bpf` related one survives today)

Imho config option should always take precedence over deprecated cmdline/env var, since otherwise falcoctl would not be able to correctly instrument the right driver for Falco.

**Comment by Andreagit97** [comment]
> I'd suggest a rebase, and then:
> 
>     * DEPRECATION NOTICE for `FALCO_BPF_PROBE`
> 
>     * DEPRECATION NOTICE for cmd line options (note that `-u` was already dropped in the meantime, so i think that only the `modern-bpf` related one survives today)
> 
> 
> Imho config option should always take precedence over deprecated cmdline/env var, since otherwise falcoctl would not be able to correctly instrument the right driver for Falco.
> 
> /cc @leogr

I agree with the proposed changes. Since we are here, we could evaluate if we should put also `gvisor` and `no-driver` in this config, in this way we have a knob for all possible engines. Moreover we need to find a way to specify additional config for each drivers, for example the bpf probe path for the `bpf` option

**Comment by FedeDP** [comment]
> Moreover we need to find a way to specify additional config for each drivers, for example the bpf probe path for the bpf option

Yep; we already have the `modern_bpf` options block; we might do the same for the other, ie: support `kmod:` and `bpf:`.
It is simple yet effective, since one can specify options for all drivers and then switch between them by just updating the `driver.type` option, without further changes to eg (non-existent): `driver.options`.

File: userspace/falco/configuration.cpp
Comment (FedeDP) [inline userspace/falco/configuration.cpp:114]: I'd call this one `modern-bpf`. 
That is how it is used today in rpm/deb packages scripts, and this is what i will use for https://github.com/falcosecurity/falcoctl/pull/343.
Also, the systemd unit is called `modern-bpf` and the CLI flag too.

File: falco.yaml
Comment (leogr) [inline falco.yaml:805]:
```suggestion
driver: 
   kind: kmod
```

I'd prefer to have the ability to include suboptions to `driver` to avoid polluting the config with first-level keys. For example:

```yaml
driver:
  kind: ebpf
  ebpf:
    # Path where the eBPF probe is located. 
    path: /root/.falco/
  modern_ebpf:
    cpus_for_each_syscall_buffer: 2
```

This will mimic [the Falco Helm chart config](https://github.com/falcosecurity/charts/blob/master/charts/falco/values.yaml#L177), which is fine IMO
 
**N.B. re. naming convention**
I believe we als have to fix the naming convention of this stuff for the last time. My proposal is as follows and include some alias to avoid headaches:

- `nodriver` (alias: `no-driver`)
- `module` (alias: `kmod`)
- `ebpf` (alias: `bpf`)
- `modern-ebpf`: (alias: `modern-bpf`)
- `gvisor` (which is technically not a driver, but I don't think we have a better option atm)

Comment (Andreagit97) [reply falco.yaml:805]: i tend to agree with this design and naming, we should follow the same approach for helm charts and systemd units

Comment (therealbobo) [reply falco.yaml:805]: I really like it! I'll update it!

Comment (Andreagit97) [reply falco.yaml:805]: when we decide something definitive, we should update this issue https://github.com/falcosecurity/falco/issues/2840: "deprecation" in Falco 0.37 means "breaking change" in Falco 0.38, so I would directly update this issue

Comment (FedeDP) [reply falco.yaml:805]: I agree with the design; about the naming, i'd go with:
* `nodriver` alias `no-driver`
* `module` alias `kmod`
* `bpf`
* `modern-bpf`
* `gvisor`

We never used `ebpf` nor `modern-ebpf` anywhere afaik. We have `FALCO_BPF_PROBE` and `--modern-bpf` today in Falco, and i'd keep them.

Falcoctl will use: `kmod`, `bpf`, `modern-bpf`.
Once again remember the naming of our systemd services too (https://github.com/falcosecurity/falco/tree/master/scripts/systemd)!

Comment (Andreagit97) [reply falco.yaml:805]: we `ebpf` in the helm chart and we tend to use `ebpf` instead of `bpf` in all the documentation https://falco.org/tags/ebpf/ because technically it is more appropriate... but I have no strong opinion here we just need to pick one

Comment (FedeDP) [reply falco.yaml:805]: Oh i see; thanks for pointing that out!

File: falco.yaml
Comment (Andreagit97) [inline falco.yaml:797]: not sure about the usage of this `custom` what would be the final use case?

File: falco.yaml
Comment (Andreagit97) [inline falco.yaml:795]: I would probably rename the `nodriver` in something else. Today the `nodriver` is a workaround to load plugins with `syscall` source, but to the final user this is not clear using a word like `nodriver`, probably something like `syscall_plugin` would be more meaningful.

File: falco.yaml
Comment (Andreagit97) [inline falco.yaml:791]: I would avoid the word `driver`. Moreover, in this list we still miss:
* the capability to read `scap-file` so something like ` mode: replay`
* the capability to open a source plugin, but this is not so easy since we have also the `load_plugin` config

File: falco.yaml
Comment (leogr) [reply falco.yaml:797]: Agree. I'd avoid introducing the `custom` option.

File: falco.yaml
Comment (leogr) [reply falco.yaml:795]: The _no driver_ option is not related to plugins. It can be helpful in case of testing or running Falco in CI.
If we believe it is misleading, we can choose a more appropriate label, still expressing that Falco will not rely on a driver. 

For example, `none` might be a good alternative:

```
driver:
  kind: none
```

Alternatively, we can use a different config key:

```
driver:
  enabled: false
```

File: falco.yaml
Comment (leogr) [reply falco.yaml:791]: I'm not totally convinced about this. The `driver` word expresses the need for an external running component to the instrument to collect the data. On the other hand, a plugin is a component that runs together with `falco` in the same process. A scap file is an entirely different use case. Moreover, I believe the `replay` mode should only be available from the command line as a flag (why one should configure it in `falco.yaml` :thinking: ?).

 Likely, your proposal tried to mimic the _scap open_ behavior. However, we should refrain from exposing how scap open works to the end users and keep the _driver_ distinct from other modes (as it is now), IMO.

File: falco.yaml
Comment (leogr) [reply falco.yaml:805]: I'm ok with picking `ebpf` as the standard one. Also, I don't see any issue with maintaining aliases.

File: falco.yaml
Comment (Andreagit97) [reply falco.yaml:795]: > The no driver option is not related to plugins. It can be helpful in case of testing or running Falco in CI.

Looking at the code it seems also related to source plugins with "syscall" source

```cpp
		else if (s.options.nodriver) /* nodriver engine. */
		{
			// when opening a capture with no driver, Falco will first check
			// if a plugin is capable of generating raw events from the libscap
			// event table (including system events), and if none is found it
			// will use the nodriver engine.
			for (const auto& p: inspector->get_plugin_manager()->plugins())
			{
				if (p->caps() & CAP_SOURCING && p->id() == 0)
				{
					auto cfg = s.plugin_configs.at(p->name());
					falco_logger::log(LOG_INFO, "Opening '" + source + "' source with plugin '" + cfg->m_name + "'");
					inspector->open_plugin(cfg->m_name, cfg->m_open_params);
					return run_result::ok();
				}
			}
			falco_logger::log(LOG_INFO, "Opening '" + source + "' source with no driver\n");
			inspector->open_nodriver();
		}
```

Comment (jasondellaluce) [reply falco.yaml:795]: The no-driver option related to plugins implementing the `syscall` raw event source. Since Falco never had a per-plugin selector for opening a capture, but rather a per-source one, the logic has always been as follows:
- When enabling the `syscall` source, it is opened by using the selected driver
- When enabling any non-`syscall` source, it sis opened by using the loaded plugin implementing it
- When having a event-sourcing plugin implementing the `syscall` source itself, the priority is still given to the selected driver for non-ambiguity and backward compatibility with the previously existing behavior. In such case, if the selected driver is `no-driver`, then the plugin is used for opening the `syscall` event source. This was introduced in Falco 0.35.0

Comment (leogr) [reply falco.yaml:795]: @jasondellaluce @Andreagit97 

My point was that `--no-driver` itself is not strictly related to plugins. On the other hand, it's related to `syscall` since the `syscall` data source needed a driver historically. The fact that we need `nodriver` for loading the event-sourcing plugin implementing the syscall source is the consequence of maintaining the backward compatibility, but we can change this behavior in the future, can't we? :thinking:

Moreover, there is no actual implementation of a syscall plugin, while `nodriver` for the CI purpose is a compelling use case currently [in use](https://github.com/falcosecurity/charts/blob/master/tests/falco-test-ci.yaml#L14C8-L14C8).


So I still think the `nodriver` option is a compelling option that deserves to exist regardless of the plugin use case. Let me know if you disagree!

File: falco.yaml
Comment (FedeDP) [reply falco.yaml:791]: I'd say: `source` instead of `driver`, and `source.mode: plugin` and deprecate `load_plugin` in next release; for now it will be supported along side (ie: `load_plugin` would override the `source.mode`.

**Comment by FedeDP** [comment]
TODO: once this is merged, we will need to update https://github.com/falcosecurity/falco/issues/2840 with new DEPRECATED options.

File: userspace/falco/app/actions/helpers_inspector.cpp
Comment (FedeDP) [inline userspace/falco/app/actions/helpers_inspector.cpp:1]: In this file we will always use config values, since CLI options have already been merged and stored in relevant config structures by `apply_deprecated_options` below.

File: userspace/falco/app/actions/load_config.cpp
Comment (FedeDP) [inline userspace/falco/app/actions/load_config.cpp:30]: Here we are merging CLI options into relevant config structures.
This way, we can trust `config` values.

File: userspace/falco/configuration.cpp
Comment (FedeDP) [inline userspace/falco/configuration.cpp:447]: We load `m_buf_size_preset` and `m_drop_failed_exit` 3 times, one for each supporting driver; this is just to avoid a switch on `m_driver_mode`.
It will be dropped in Falco 0.38 anyway.

**Comment by FedeDP** [comment]
This PR is quite ready; we are just choosing a better name than `driver` for the new config section :)

**Comment by FedeDP** [comment]
Rebased on top of master.

**Comment by FedeDP** [comment]
Renamed `driver.` to `engine.` in config; morever, renamed `engine.replay.scap_file` to `engine.replay.trace_file`.
This is ready.
/cc @leogr

File: userspace/falco/app/actions/configure_syscall_buffer_num.cpp
Comment (Andreagit97) [reply userspace/falco/app/actions/configure_syscall_buffer_num.cpp]: I should have solved all these issues in the commits I added :)

**Comment by FedeDP** [comment]
Great catch Andre, thanks!

**Review by Andreagit97 — DISMISSED** [review]
This LGTM!
/approve

**Review by Andreagit97 — COMMENTED** [review]
LGMT!

File: userspace/falco/app/actions/helpers_inspector.cpp
Comment (leogr) [inline userspace/falco/app/actions/helpers_inspector.cpp:39]:
```suggestion
		falco_logger::log(falco_logger::level::INFO, "Replaying events from the capture file: " + s.config->m_replay.m_capture_file + "\n");
```

N.B. _capture files_ are not tied to system call

File: userspace/falco/app/actions/load_config.cpp
Comment (leogr) [inline userspace/falco/app/actions/load_config.cpp:36]: I'm fine with this behavior. However, I don't like silently ignoring deprecated options here (users may not get what's happening).

So, when the user mistakenly uses _BOTH_ the command line flag _AND_ change the config, we should print something like:

_The '-g,--gvisor-config' command line option is deprecated and will be removed in Falco version 0.38. Note that this option is ineffective when 'engine' settings are specified via the configuration file._

File: userspace/falco/app/actions/load_config.cpp
Comment (leogr) [inline userspace/falco/app/actions/load_config.cpp:65]:
```suggestion
		falco_logger::log(falco_logger::level::WARNING, "DEPRECATION NOTICE: the '-g,--gvisor-config' command line option is deprecated and will be removed in Falco 0.38!\n");
```

just nit :)

File: userspace/falco/app/actions/load_config.cpp
Comment (leogr) [inline userspace/falco/app/actions/load_config.cpp:47]: What will happen when [this](https://github.com/falcosecurity/falco/pull/2918) gets merged, and perhaps some of the users will use the same env var in the config :point_down: ? 

```yaml
engine:
  kind: kmod
  kmod:
    buf_size_preset: 4
    drop_failed_exit: false
  ebpf:
    # path to the elf file to load.
    probe: ${FALCO_BPF_PROBE}
```

:exploding_head:

File: userspace/falco/configuration.cpp
Comment (leogr) [inline userspace/falco/configuration.cpp:119]: These defaults are a bit hidden. However, not a big issue.

File: userspace/falco/configuration.h
Comment (leogr) [inline userspace/falco/configuration.h:47]: Do you know if this is a particular reason for this ordering?

Not saying it is an issue, just I don't get it.

**Review by leogr — COMMENTED** [review]
Love this. It's a huge improvement, and we will finally get a bit of consistency across our configuration. So, first of all, a big thanks to @therealbobo, @Andreagit97, and @FedeDP :clap: 

The implementation is already good. I've just left some comments about minor issues I've found when reviewing.

Other than that, I want to ensure that a warning message is printed to the user any time a deprecated option is used (either command line or config). I've already identified a corner case (see my comment below) and I think another similar corner case can happen if the user modified both the old and the new field in the configuration file.

Ping me privately, I can provide more details if needed.

File: userspace/falco/configuration.h
Comment (FedeDP) [reply userspace/falco/configuration.h:47]: No special reason!

File: userspace/falco/configuration.cpp
Comment (FedeDP) [reply userspace/falco/configuration.cpp:119]: Yep, they will go away in next release cycle.

File: userspace/falco/app/actions/load_config.cpp
Comment (FedeDP) [reply userspace/falco/app/actions/load_config.cpp:36]: I'd add a single log, something like:
```
if(s.config->m_changes_in_engine_config)
	{
                   falco_logger::log(falco_logger::level::WARNING, "Since you are using the new engine configuration key, deprecated command line options [-e,-g,...] and FALCO_BPF_PROBE env variable will be ignored.\n");
		return run_result::ok(); 
```

File: falco.yaml
Comment (FedeDP) [reply falco.yaml:330]: > I would make Falco output a clear error message instructing the user to pass a value to replay.capture_file

There is already one. Agree about the default!

File: falco.yaml
Comment (FedeDP) [reply falco.yaml:334]: I'd leave it without any example, we can iterate it further in a follow up PR if needed.

File: userspace/falco/app/actions/load_config.cpp
Comment (FedeDP) [reply userspace/falco/app/actions/load_config.cpp:39]: Added in latest commit; example output:
```
Thu Nov 23 15:15:50 2023: Falco version: 0.37.0-146+9bf0565 (x86_64)
Thu Nov 23 15:15:50 2023: Falco initialized with configuration file: ../falco.yaml
Thu Nov 23 15:15:50 2023: Since the new 'engine' config key is being used, deprecated CLI options [-e,-g,--gvisor-config,--nodriver,--modern-bpf] and FALCO_BPF_PROBE environment variable will be ignored.
Thu Nov 23 15:15:50 2023: Loading rules from file ../rules/falco_rules.yaml
```

File: userspace/falco/app/actions/load_config.cpp
Comment (leogr) [reply userspace/falco/app/actions/load_config.cpp:50]: > Nothing, i mean the deprecated env variable takes precedence.

Will it print a warning even if it shouldn't? :thinking:

Comment (FedeDP) [reply userspace/falco/app/actions/load_config.cpp:50]: Yes it will print the newly introduced warning about the fact that FALCO_BPF_PROBE env variable will be ignored, and that is correct(since we won't directly use the env variable but the new config key that resolves to the env variable value)

File: userspace/falco/app/actions/load_config.cpp
Comment (leogr) [reply userspace/falco/app/actions/load_config.cpp:39]: Ok. I was thinking of printing the warning only if one of those options was used. But your solution is fine anyway :+1:

File: userspace/falco/app/actions/load_config.cpp
Comment (FedeDP) [reply userspace/falco/app/actions/load_config.cpp:50]: Oh i mis-read (i thought `kind: bpf` was in use). In this case, as i stated earlier, the deprecated env variable takes precedence (since the `bpf` kind configuration is not even parsed, because the `kind` is kmod).
No warning in this case (except for the `FALCO_BPF_PROBE` env var deprecation )

**Review by LucaGuerra — APPROVED** [review]
Thank you for this! 😮 LGTM, so we can test it better

**Review by leogr — APPROVED** [review]
For sure we need to deeply test it, but LGTM at the moment!

Thank you'll!

**Pull request merged**
